src/optimal_image/search.rs

```diff
@@ -1,37 +1,132 @@
-        self
+        let extension = path.extension().unwrap().to_str();
+
+        let original;
+        if extension == Some("jpg") {
+            let dataclient = Jpeg::new();
+            original = dataclient.load(path)?;
+        } else if extension == Some("png") {
+            let dataclient = Png::new();
+            original = dataclient.load(path)?;
+        } else {
+            return Err(Box::new(ImageError::FormatError(String::from(
+                "error",
+            ))));
+        }
+
+        let search = Search {
+            path: path.to_str().unwrap().to_string(),
+            image_data: original,
+            options: options,
+            result: None,
+        };
+
+        Ok(search)
+    }
+
+    pub fn run(&mut self) -> Option<SearchResult> {
+        // let mut config_map: HashMap<&ImageFormat, Vec<ImageConfig>> = HashMap::new();
+        let configs = self.options.formats.iter().fold(
+            HashMap::new(),
+            |mut map: HashMap<&ImageFormat, Vec<ImageConfig>>, format| {
+                let conf_list = (self.options.quality_range.0
+                    ..self.options.quality_range.1)
+                    .map(|q| {
+                        let fmt = format.clone();
+                        ImageConfig {
+                            id: self.path.clone(),
+                            encoding_config: EncodingConfig {
+                                format: fmt,
+                                quality: q,
+                            },
+                        }
+                    })
+                    .collect();
+
+                map.insert(format, conf_list);
+                map
+            },
+        );
+
+        let result: SearchResult = configs.iter().fold(
+            HashMap::new(),
+            |mut result: SearchResult, (&k, v)| {
+                let optimal_config = find_optimal_config(
+                    &self,
+                    &v,
+                    self.options.threshold,
+                );
+                result.insert(k.clone(), optimal_config.clone());
+                result
+            },
+        );
+
+        println!("{:?}", result);
+
+        self.result = Some(result.clone());
+
+        Some(result)
+    }
+
+    pub fn get_result(&mut self) -> &Option<SearchResult> {
+        &self.result
     }
 }
 
-pub fn find_optimal_config(
-    config: &Vec<ImageConfig>,
+pub fn find_optimal_config<'a>(
+    search: &Search,
+    config: &'a &Vec<ImageConfig>,
     threshold: f64,
-) -> &ImageConfig {
+) -> &'a ImageConfig {
+    let image_jpeg_image = ImageJpegImage::from_image_data(&search.image_data).unwrap();
     let index = config
         .binary_search_by(|probe| {
-            // TODO:
-            // generate image with `probe` config
-            // run dssim against original
-            // return Ordering according to threshold
-            unimplemented!();
+            let mut dssim_context = dssim::new();
```

Comment:
this could eventually be persisted, but is quite hard to make the borrow checker happy with a mut ref